File: spare-parts-shop/src/main/java/com/spareparts/model/EMIInstallment.java

package com.spareparts.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "emi_installments")
@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class EMIInstallment implements BelongsToBusiness {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "business_id", nullable = false)
    @JsonIgnore
    private Business business;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "branch_id")
    private Branch branch;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "emi_id", nullable = false)
    private EMI emi;

    @Column(name = "installment_number", nullable = false)
    private Integer installmentNumber;

    @Column(name = "due_date", nullable = false)
    private LocalDate dueDate;

    @Column(name = "amount", nullable = false)
    private Double amount;

    @Column(name = "paid_amount")
    private Double paidAmount = 0.0;

    @Column(name = "remaining_amount")
    private Double remainingAmount;

    @Column(name = "status", nullable = false)
    private String status = "PENDING"; // PENDING, PAID, OVERDUE, PARTIALLY_PAID, CANCELLED

    @Column(name = "paid_date")
    private LocalDate paidDate;

    @Column(name = "payment_method")
    private String paymentMethod;

    @Column(name = "late_fee")
    private Double lateFee = 0.0;

    @Column(name = "remarks")
    private String remarks;

    @Column(name = "created_at")
    private LocalDateTime createdAt = LocalDateTime.now();

    @PrePersist
    protected void prePersist() {
        if (this.remainingAmount == null) {
            this.remainingAmount = this.amount;
        }
    }
}

File: spare_parts_shop_flutter/lib/models/emi_installment.dart

class EMIInstallment {
  final int id;
  final int emiId;
  final int installmentNumber;
  final String dueDate;
  final double amount;
  final String status;
  final String? paidDate;
  final String? paymentMethod;
  final double lateFee;
  final String? remarks;
  final String createdAt;

  EMIInstallment({
    required this.id,
    required this.emiId,
    required this.installmentNumber,
    required this.dueDate,
    required this.amount,
    required this.status,
    this.paidDate,
    this.paymentMethod,
    required this.lateFee,
    this.remarks,
    required this.createdAt,
  });

  factory EMIInstallment.fromJson(Map<String, dynamic> json) {
    return EMIInstallment(
      id: json['id'],
      emiId: json['emi']['id'],
      installmentNumber: json['installmentNumber'],
      dueDate: json['dueDate'],
      amount: (json['amount'] as num).toDouble(),
      status: json['status'],
      paidDate: json['paidDate'],
      paymentMethod: json['paymentMethod'],
      lateFee: (json['lateFee'] as num).toDouble(),
      remarks: json['remarks'],
      createdAt: json['createdAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'emiId': emiId,
      'installmentNumber': installmentNumber,
      'dueDate': dueDate,
      'amount': amount,
      'status': status,
      'paidDate': paidDate,
      'paymentMethod': paymentMethod,
      'lateFee': lateFee,
      'remarks': remarks,
      'createdAt': createdAt,
    };
  }
}
